presto-main/src/main/java/com/facebook/presto/operator/scalar/KleinbergBurstDetectionFunction.java

```diff
@@ -0,0 +1,212 @@
+        double[] alphaLn = new double[burstLevelUpperLimit];
+
+        alpha[0] = numEntries / timeLength;  // Average event rate.
+        alphaLn[0] = Math.log(alpha[0]);
+
+        for (int i = 1; i < burstLevelUpperLimit; ++i) {
+            alpha[i] = s * alpha[i - 1];
+            alphaLn[i] = Math.log(alpha[i]);
+        }
+
+        // Initialize.
+        Node[] q = new Node[burstLevelUpperLimit];
+        double[] costs = new double[burstLevelUpperLimit];
+        Arrays.fill(costs, Double.MAX_VALUE);
+        costs[0] = 0;
+
+        // Start optimization.
+        double gammalogn = gamma * Math.log((double) numEntries);
+        for (double interval : intervals) {
+            Node[] qNew = new Node[burstLevelUpperLimit];
+            double[] costsNew = new double[burstLevelUpperLimit];
+            for (int i = 0; i < burstLevelUpperLimit; ++i) {
+                double[] c = new double[burstLevelUpperLimit];
+                for (int j = 0; j < burstLevelUpperLimit; ++j) {
+                    c[j] = costs[j] + tau(gammalogn, j, i);
+                }
+                int jMin = findSmallestIndex(c);
+                // Store the cost for setting the burst level i.
+                costsNew[i] = -logf(alpha, alphaLn, i, interval) + c[jMin];
+                qNew[i] = new Node(i, q[jMin]);
+            }
+            q = qNew;
+            costs = costsNew;
+        }
+
+        // Build output structure.
+        List<Integer> bursts = new ArrayList<Integer>(Collections.nCopies(numEntries, 0));
+        int seqMin = findSmallestIndex(costs);
+        int count = 0;
+        Node p = q[seqMin];
+        while (p != null) {
+            bursts.set(numEntries - ++count, p.state);
+            p = p.parent;
+        }
+
+        BlockBuilder burstsBuilder =
+                IntegerType.INTEGER.createBlockBuilder((BlockBuilderStatus) null, numEntries);
+        for (int i = 0; i < numEntries; i++) {
+            BlockBuilder burstItem = IntegerType.INTEGER.createBlockBuilder((BlockBuilderStatus) null, 1);
+            IntegerType.INTEGER.writeLong(burstItem, bursts.get(i));
+            IntegerType.INTEGER.appendTo(burstItem, 0, burstsBuilder);
+        }
+
+        return burstsBuilder.build();
+    }
+
+    // Internal helper structure to hold discovered burst nodes.
+    private class Node
+    {
+        public int state; // burst level: 0, 1, ..., K-1
+        public Node parent;
+
+        public Node(int s, Node parent)
+        {
+            this.state = s;
+            this.parent = parent;
+        }
+    };
+
+    /**
+     * Cost function for reproducing a given interval.
+     * 'i': the burst level
+     * 'x': interval
+     */
+    private static double logf(double[] alpha, double[] lnAlpha, int i, double x)
+    {
+        return lnAlpha[i] - alpha[i] * x;
+    }
+
+    /**
+     * Cost function for changing a burst level.
+     * 'i': the previous burst level
```

Comment:
Javadoc should use `@param` for these. See http://www.oracle.com/technetwork/java/javase/tech/index-137868.html